snarkos/ledger/mod.rs

```diff
@@ -25,36 +37,81 @@ use std::{
     net::{IpAddr, SocketAddr},
     sync::Arc,
 };
-use tokio::task;
+use tokio::{
+    sync::{mpsc, oneshot},
+    task,
+};
 use warp::{reply, Filter, Rejection, Reply};
 
+/// Shorthand for the parent half of the `Ledger` message channel.
+pub(crate) type LedgerRouter<N> = mpsc::Sender<LedgerRequest<N>>;
+#[allow(unused)]
+/// Shorthand for the child half of the `Ledger` message channel.
+type LedgerHandler<N> = mpsc::Receiver<LedgerRequest<N>>;
+
+///
+/// An enum of requests that the `Ledger` struct processes.
+///
+#[derive(Debug)]
+pub enum LedgerRequest<N: Network> {
+    /// BlockResponse := (peer_ip, block)
+    BlockResponse(SocketAddr, Block<N>),
+    /// BlockRequest := (peer_ip, block_height)
+    BlockRequest(SocketAddr, u32),
+    /// Disconnect := (peer_ip, reason)
+    Disconnect(SocketAddr, DisconnectReason),
+    /// Failure := (peer_ip, failure)
+    Failure(SocketAddr, String),
+    /// Heartbeat
+    Heartbeat,
+    /// Ping := (peer_ip)
+    Ping(SocketAddr),
+    /// Pong := (peer_ip, block_height, node_type)
+    Pong(SocketAddr, u32, Option<NodeType>),
+    /// UnconfirmedBlock := (peer_ip, block)
+    UnconfirmedBlock(SocketAddr, Block<N>),
+    /// UnconfirmedTransaction := (peer_ip, transaction)
+    UnconfirmedTransaction(SocketAddr, Transaction<N>),
+}
+
 pub(crate) type InternalLedger<N> = snarkvm::prelude::Ledger<N, BlockDB<N>, ProgramDB<N>>;
 // pub(crate) type InternalLedger<N> = snarkvm::prelude::Ledger<N, BlockMemory<N>, ProgramMemory<N>>;
 
 pub(crate) type InternalServer<N> = snarkvm::prelude::Server<N, BlockDB<N>, ProgramDB<N>>;
 // pub(crate) type InternalServer<N> = snarkvm::prelude::Server<N, BlockMemory<N>, ProgramMemory<N>>;
 
-pub(crate) type Peers<N> = Arc<RwLock<IndexMap<SocketAddr, crate::Sender<N>>>>;
+pub type PeersState = IndexMap<SocketAddr, Option<NodeType>>;
 
 #[allow(dead_code)]
-pub struct Ledger<N: Network> {
+pub struct Ledger<N: Network, E: Environment> {
```

Comment:
We may want to consider not routing through the ledger in future. This would reduce complexity. I'll suggest something alongside the upcoming kadmium + noise additions.